src/librustc_typeck/astconv.rs

```diff
@@ -1247,8 +1247,10 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
         // Expand trait aliases recursively and check that only one regular (non-auto) trait
         // is used and no 'maybe' bounds are used.
         let expanded_traits =
-            traits::expand_trait_aliases(tcx, bounds.trait_bounds.iter().cloned());
-        let (mut auto_traits, regular_traits): (Vec<_>, Vec<_>) =
+            traits::expand_trait_aliases(tcx, bounds.trait_bounds.iter().cloned())
+                // Ensure that trait ref is to self type and not some type param.
```

Comment:
I'm a bit worried here about an example like this:

```rust
#![feature(trait_alias)]

trait Foo<T> = where T: PartialEq<Self>;

fn main() {
    let _: &dyn Foo<()> = &();
}
```

When I test this, I currently get some wacky ICE =) But it seems like we would screen out the `T: PartialEq<Self>` from the list of `regular_traits` here. Then, later, when we check for object safety, we will never see that `T: PartialEq<Self>` is not object-safe. 

My intuition is that `trait Foo<T> = ...` should work exactly as if you created a trait and implemented it, for the most part. And [if we do that](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=8dab4b2fc3984d04ee5f38f5b0c87f5d) with this example we get object safety errors. I don't quite see where they are supposed to crop up here.

Reply:
The key thing is to avoid trying to create existential predicates for predicates where the self ty is not the dummy self ty, which is of course impossible. This filter might be blocking out *too* much (projections on `Self` too?), but all the tests seem to be working anyway...

The relevant test is `issue-65673.rs`, I believe (pre-existing, slightly modified).

Reply:
`trait Foo = where Self: Display;` should (and I think is) equivalent to `trait Foo = Display;` in most circumstances, except as a trait object. Since technically no principal trait is specified, the compiler will complain. Perhaps we should do that, but perhaps we should instead check that there's at least one predicate instead of trait?

Reply:
As for object safety, I guess we need to modify that code explicitly? Given such predicates on `T` definitely cannot be included in the trait object itself (they're simply not existential).

Reply:
I don't quite understand your response here, @alexreg. 

What is the behavior on the example I gave, for example? I don't think that `issue-65673.rs` is equivalent. For one thing, it's run-pass, but I expect an error, but also the trait alias in `issue-65673.rs` is:

```rust
 trait Alias<T> = Any where T: Trait; 
```

whereas the trait alias in my example was `trait Foo<T> = where T: PartialEq<Self>;`, which (notably) references `Self` (unlike `issue-65673.rs`).

Also, regarding the difference between `trait Foo = Display` and `trait Foo where Self: Display`, I guess I'd have to consult the RFC. It's a *touch* surprising to me since `trait Foo: Display` and `trait Foo where Self: Display` *are* equivalent, for better or worse.

Reply:
@nikomatsakis Okay, I've gotten your expected behaviour implemented. (And I managed to remove the `Any` from `issue-65673.rs`.) Now, the question is, do we want to allow just `dyn` or `dyn 'static` as a type?

Reply:
I don't know! I think I'm inclined to preserve the status quo, whatever it is, though if/when we generalize to `dyn A + B` I'd also be inclined to accept zero bounds.

Reply:
@nikomatsakis Okay, sounds fair. I'll push shortly and this PR will be ready for review.